loopd/daemon.go

```diff
@@ -119,117 +135,291 @@ func daemon(config *Config, lisCfg *listenerCfg) error {
+		return errOnlyStartOnce
+	}
+
+	// When starting as a subserver, we get passed in an already established
+	// connection to lnd that might be shared among other subservers.
+	d.lnd = lndGrpc
+
+	// With lnd already pre-connected, initialize everything else, such as
+	// the swap server client, the RPC server instance and our main swap
+	// handlers.
+	err := d.initialize()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
 
+// initialize creates and initializes an instance of the swap server client,
+// the swap client RPC server instance and our main swap handlers.
+func (d *Daemon) initialize() error {
+	// If no swap server is specified, use the default addresses for mainnet
+	// and testnet.
+	if d.cfg.SwapServer == "" {
+		// TODO(wilmer): Use onion service addresses when proxy is
+		// active.
+		switch d.cfg.Network {
+		case "mainnet":
+			d.cfg.SwapServer = mainnetServer
+		case "testnet":
+			d.cfg.SwapServer = testnetServer
+		default:
+			return errors.New("no swap server address specified")
+		}
+	}
+
+	log.Infof("Swap server address: %v", d.cfg.SwapServer)
+
+	// Create an instance of the loop client library.
+	swapclient, clientCleanup, err := getClient(d.cfg, &d.lnd.LndServices)
+	if err != nil {
+		return err
+	}
+	d.clientCleanup = clientCleanup
+
+	// Both the client RPC server and and the swap server client should
+	// stop on main context cancel. So we create it early and pass it down.
+	d.mainCtx, d.mainCtxCancel = context.WithCancel(context.Background())
+
+	// Now finally fully initialize the swap client RPC server instance.
+	d.swapClientServer = swapClientServer{
+		impl:        swapclient,
+		lnd:         &d.lnd.LndServices,
+		swaps:       make(map[lntypes.Hash]loop.SwapInfo),
+		subscribers: make(map[int]chan<- interface{}),
+		statusChan:  make(chan loop.SwapInfo),
+		mainCtx:     d.mainCtx,
+	}
+
+	// Retrieve all currently existing swaps from the database.
+	swapsList, err := d.impl.FetchSwaps()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range swapsList {
+		d.swaps[s.SwapHash] = *s
+	}
+
+	// Start the main event handlers, now that we have prepared everything
+	// they need to work.
+	d.startSwapHandlers()
+
+	return nil
+}
+
+// startSwapHandlers starts all goroutines that are tasked with handling swap
+// events.
+func (d *Daemon) startSwapHandlers() {
 	// Start the swap client itself.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
-
 		log.Infof("Starting swap client")
-		err := swapClient.Run(mainCtx, server.statusChan)
+		err := d.impl.Run(d.mainCtx, d.statusChan)
 		if err != nil {
+			// Only log the error here for now. We'll send it over
+			// the errChan later, after we've stopped the daemon.
 			log.Error(err)
 		}
 		log.Infof("Swap client stopped")
 
-		log.Infof("Stopping gRPC server")
-		grpcServer.Stop()
+		// We need to mark this goroutine to be done explicitly here.
+		// Otherwise the stop below will wait for the defer which will
+		// end in a deadlock.
```

Comment:
To me this comment signals that we are not on the optimal design yet. There are currently four goroutines running: rest, grpc, swap client and status updates. Maybe a main goroutine is needed to listen on all error channels and coordinate the stopping process.

Reply:
Yes, that makes sense, I see it now. It does look cleaner that way.